src/child-applications/lifecycles/unload.js

```diff
@@ -13,23 +13,34 @@ export async function toUnloadPromise(app) {
 		 * anything that still thinks we need to unload the app.
 		 */
 		finishUnloadingApp(app, unloadInfo);
-		return;
+		return app;
+	}
+
+	if (app.status === UNLOADING) {
+		/* Both unloadChildApplication and reroute want to unload this app.
+		 * It only needs to be done once, though.
+		 */
+		await unloadInfo.promise;
+		return app;
 	}
 
-	if (app.status !== NOT_MOUNTED || !unloadInfo) {
-		/* Either no one has called unloadChildApplication for this app,
-		 * or it's not in a status where it can be unloaded.
+	if (app.status !== NOT_MOUNTED) {
+		/* The app cannot be unloaded until it is unmounted.
 		 */
-		return;
+		return app;
+	}
+	
+	if (!unloadInfo) {
+		/* No one has called unloadChildApplication for this app,
+		 */
+		return app;
 	}
 
 	try {
 		app.status = UNLOADING;
 		await reasonableTime(app.unload(), `Unloading application '${app.name}'`, app.timeouts.unload);
 	} catch (err) {
-		handleChildAppError(err, app);
-		app.status = SKIP_BECAUSE_BROKEN;
-		unloadInfo.reject(err);
+		errorUnloadingApp(app, unloadInfo, err);
```

Comment:
Another bug I found because of the new events-api tests. We were not actually cleaning up apps when they broke during unload, which resulted in every reroute thinking that it needed to re-unload the broken app.